### DISCUSSION TITLE
Usecases of the "Recover RBD/CephFS RWO PVC in case of Node Loss" feature

### DISCUSSION THREAD

**Discussion by cupnes** [rook/rook#14311] [General]
I am currently verifying the behavior of RBD during node failure. And I found the "Recover RBD/CephFS RWO PVC in case of Node Loss" feature. I have a question about usecases of this feature.

This feature assumes that a pod using RBD PVC is running on the node at the time the taint out-of-service is added to the node.

https://github.com/rook/rook/blob/master/design/ceph/node-loss-rbd-cephfs.md
> Once this taint is applied manually, Rook will create a NetworkFence CR. The csi-addons operator will then blocklist the node to prevent any ceph rbd/CephFS client on the lost node from writing any more data.

However, a K8s official document describes the usage of `out-of-service` taint as follows.

https://kubernetes.io/blog/2022/12/16/kubernetes-1-26-non-graceful-node-shutdown-beta/
> Note: Before applying the out-of-service taint, you must verify that a node is already in shutdown or power-off state (not in the middle of restarting).

If `out-of-service` taint is used as written in this document, NetworkFence seems not to be necessary.

Could you tell me the usecases of this feature? Is there any case where `out-of-service` is set even if the target node is shutdown?

**Comment by travisn** [comment]
The NetworkFence is necessary to inform the Ceph RBD volume that it is safe to detach, and that the node should be fenced in case the node comes back online unexpectedly. The RBD volume cannot be allowed to come back online from that node, since it would potentially cause corruption.

**Comment by cupnes** [comment]
@travisn 
Thank you for your response.

> The NetworkFence is necessary to inform the Ceph RBD volume that it is safe to detach, and that the node should be fenced in case the node comes back online unexpectedly. The RBD volume cannot be allowed to come back online from that node, since it would potentially cause corruption.

Does it mean there are cases that `out-of-service` taint might be set when the node becomes network unreachable (in other words, the node is not in the shutdown state)? If so, I have one more question. There is a race between the creation of NetworkFence by Rook and the deletion of VolumeAttachment by Non-graceful shutdown. If these two processes happen in order, it works fine. However, I'm afraid of the following case.

1. There is a pod P0 mounting an RBD PVC R0 on top of node N0.
2. N0 becomes network unreachable.
3. An admin set `out-of-service` to N0.
4. K8s removes the VolumeAttachment about N0 and R0 thanks to Non-graceful shutdown.
5. Another pod P1 on N1 mounts R0.
6. Both P0 and P1 write data to R0 and it results in data corruption. 
7. Rook creates the NetworkFence about N0.

What do you think?

**Reply by travisn** [reply]
@subhamkrai Please review this for network fencing

**Reply by cupnes** [reply]
@subhamkrai How is the status of the review?

**Reply by subhamkrai** [reply]
@cupnes we do fencing for RWO, so step 5 becomes invalid

**Reply by cupnes** [reply]
@subhamkrai If the Pod mounting the RBD PVC is created by a StatefulSet, I think that's the case. For Pods created by a Deployment, I think step 5 could occur even with RWO. What do you think?

**Reply by subhamkrai** [reply]
By another pod I thought it's whole different pod. Once you set `out-of-service` taint pods on that node will go in terminating state so there is no pod P0 also rook will do networkFence so the node with `out-of-service`  will not be reachable to ceph as networkFence do blocklisting

**Reply by cupnes** [reply]
@subhamkrai 
> By another pod I thought it's whole different pod.

My previous explanation was insufficient, so let me clarify. P0 and P1 are related to a StatefulSet or a Deployment. P1 is created as an alternative to P0 as a result of the termination of P0. With a StatefulSet, P1 is created after P0 is deleted, which should work fine. However, with a Deployment, P1 might be created while P0 is still terminating(*1), and P0 might access R0 during this state.

(*1) https://kubernetes.io/docs/concepts/workloads/controllers/deployment/
> If you manually delete a Pod, the lifecycle is controlled by the ReplicaSet and the replacement will be created immediately (even if the old Pod is still in a Terminating state). If you need an "at most" guarantee for your Pods, you should consider using a StatefulSet.

> Once you set out-of-service taint pods on that node will go in terminating state so there is no pod P0
> also rook will do networkFence so the node with out-of-service will not be reachable to ceph as networkFence do blocklisting

P0 might be running because the deletion of P0 is not guaranteed here as I described above. In addition, it is not also guaranteed that Rook's networkFence happens before the creation of P1.

**Reply by subhamkrai** [reply]
Its Kubernetes features of adding the taint `out-of-service` to move resources to different node, the moment you add this taint the pod goes to terminating state and the backend Volumeattachement mentioned in the mode is release(and it is quite fast what I have noticed) and also p1 will take his time to get running state and also mean while rook is doing fencing in the backend.

What you are saying is very very very corner and worst I guess and as you mentioned there are lots of ifs and buts.

cc @Madhu-1

**Reply by BlaineEXE** [reply]
I believe that the case being described is the reason why network fencing is needed -- and the fundamental thing that our network fencing implementation helps guard against.

If node N0 truly died, then a network fence isn't all that necessary because all clients are full dead. hen the node comes back online, there will be no old Ceph client mounts active on the node, so any attempt by CSI to mount RWO mounts will fail, due to the mount being active elsewhere.

But if node N0 has a network failure, then fencing is critical. Pods on N0 can't be stopped because the K8s control plane can't instruct kubelet on N0 to do anything. New versions of P0 and R0 (call them P0' and R0') get scheduled on a new node. The Ceph-CSI driver on N0 has to be fenced because once the network on N0 resumes, R0 and R0' are both connected again, and only R0' should be active

However, this is also why we have implemented fencing that coordinates with CSI. I think @Madhu-1 will have some more nuance or clarity, but personally, I don't think the issue @cupnes is worried about sounds like it's beyond what we have implemented. I think it directly within what we have implemented.

That's a very long-winded way of explaining everything underneath what Travis has already said here: https://github.com/rook/rook/discussions/14311#discussioncomment-9704182

**Reply by Madhu-1** [reply]
Yes, @BlaineEXE and @subhamkrai covered all the things from Rook's side. `out-of-service` is a way to tell Rook that fencing is required because we don't know whether the node is dead or not, we want to make use of the kubernetes feature to remove few things like VA and move pod around and also at the same time we want to ensure that the data inconsistency should also not happen, This is an extra guard block client I/O access from specific nodes.

**Reply by cupnes** [reply]
@subhamkrai @BlaineEXE @Madhu-1 
Thank you very much for your detailed response. I understood it as the following:

- If the node with the `out-of-service` taint is shut down, there is no issue.
- If it is not shut down, theoretically, there is a possibility that the VolumeAttachment could be deleted before fencing, allowing a terminating pod's replacement pod to start running. However, since the fencing process is faster than the startup of the replacement pod, this does not practically pose a problem.

@subhamkrai 
Your repeated and detailed responses have been incredibly helpful. I am truly grateful for your assistance. Thank you very much.

**Reply by subhamkrai** [reply]
@cupnes is this ready to close?

**Reply by cupnes** [reply]
@subhamkrai Yes. I will close this issue. Thank you.

**Comment by reefland** [comment]
Q: when a node comes back up but still has the network fence applied, what is the expectation of the OSD(s) that the node hosts?  It's not clear from the docs if the OSD recovery can happen while the taints/fence still in place.

Do the OSD(s) need to remain down? is a custom toleration needed to allow OSD recovery to start with taints/fence still in place?

**Reply by subhamkrai** [reply]
if nodes come back, admin/user need to remove the taint `out-of-service`  so that the nodes is readable by Kubernetes, and anyway all the pods will move to new node including osd if the node is available so OSD recovery should not take extra effort.

**Reply by reefland** [reply]
To clarify my question.... is there a reason Rook does not have a toleration to the `out-of-service` taint, to allow Rook to recover the hosts' OSDs when a node comes back on-line??  I thought the  `out-of-service` taint and network fence is to keep clients out, but why keep Rook out as well?

**Reply by subhamkrai** [reply]
I'm not sure if I understand correctly what you mean by `why keep Rook out as well`?

**Reply by BlaineEXE** [reply]
I believe the question is why the storage system (Rook+Ceph) is being kept out of service with the same priority and criteria as user applications (`out-of-service` taint).

I think the question is a worthwhile one to ask, since the storage system is usually a higher priority than user applications, and to some admins might view storage as being as important as the system itself. However, key words here are "usually" and "some" -- not all.

Kubernetes' own blog states that the `out-of-service` feature is intended for non-graceful node shutdown events. Kubernetes expects admins to set this on a node when it is powered off non-gracefully, and then remove it when the node is powered back on. https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/

With that as a baseline, I think Rook is doing the right thing by keeping itself out of service on a node when `out-of-service` is set. Node shutdown/crash is more fundamental than the software-defined storage that runs on the node.

It is also important for Rook to not bypass important maintenance operations that administrators need to use. Rook has to draw boundaries around what it can/can't do and will/won't do so that users can have clear ways of interacting with it.

If an administrator wants to allow Rook to come online on a previously-`out-of-service` node before user applications, there is a tool already available for that: custom taints. A user can add a taint like `my.org/system-apps-only` to a node that Rook can tolerate but user apps cannot. That would allow nuanced and tiered startup without breaking the assumptions that many users will have of the `out-of-service` tool.
